src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -307,6 +318,54 @@ namespace
         type_indexes.erase(TypeIndex::UInt64);
     }
 
+    /// if setting 'try_infer_variant' is true then we convert to type variant.
+    void transformVariant(DataTypes & data_types, TypeIndexesSet & type_indexes)
+    {
+        auto typesAreEqual = checkIfTypesAreEqual(data_types);
+        auto typesContainVariant = checkIfTypesContainVariant(data_types);
+        if (typesAreEqual)
+            return;
+
+        DataTypes new_data_types;
+        TypeIndexesSet new_type_indexes;
+        std::shared_ptr<DataTypeVariant> variant_type;
+
+        /// extract the nested types of variant and make a new variant with the nested types and the other type.
+        /// eg. Type 1: variant<String, Array>, Type 2: Date -> variant<String, Array, Date>.
+        if (typesContainVariant)
```

Comment:
```suggestion
        if (checkIfTypesContainVariant(data_types))
```

Reply:
Actually, I think we don't need this check `checkIfTypesContainVariant` and 2 branches. We can do it in one loop. Also we could have more than 1 Variant type and we should handle it:
```c++
DataTypes variant_types;
for (const auto & type : data_types)
{
    if (const auto * variant_type = typeid_cast<const DataTypeVariant *>(type.get()))
    {
        const auto & current_variants = variant_type->getVariants();
        variant_types.insert(variant_types.end(), current_variants.begin(), current_variants.end())
    }
    else
    {
        variant_types.push_back(type);
    }
}

variant_type = std::make_shared<DataTypeVariant>(variant_types);
```